pkg/controller/external_vpc.go

```diff
@@ -0,0 +1,127 @@
+		return
+	}
+	lRouters := make(map[string]util.LogicalRouter)
+	if len(extLrs) == 0 {
+		klog.V(4).Info("sync over, no external vpc")
+	}
+
+	for _, aLr := range extLrs {
+		var aRouter util.LogicalRouter
+		aRouter.Name = aLr[Name][0]
+		var ports []util.Port
+		for _, portId := range aLr[Port] {
+			port, err := c.ovnClient.GetEntityInfo("logical_router_port", portId, []string{"name"})
+			if err != nil {
+				klog.Info("get port error")
+				continue
+			}
+			aPort := util.Port{
+				Name:   port["name"],
+				Subnet: "",
+			}
+			ports = append(ports, aPort)
+		}
+		aRouter.Ports = ports
+		lRouters[aRouter.Name] = aRouter
+	}
+	UUID := "_uuid"
+	for k, v := range lRouters {
+		for _, port := range v.Ports {
+			peerPorts, err := c.ovnClient.CustomFindEntity("logical_switch_port", []string{UUID}, fmt.Sprintf("options:router-port=%s", port.Name))
+			if err != nil || len(peerPorts) > 1 {
+				klog.Errorf("failed to list peer port of %s, %v", port, err)
+				continue
+			}
+			if len(peerPorts) == 0 {
+				continue
+			}
+			switches, err := c.ovnClient.CustomFindEntity("logical_switch", []string{Name}, fmt.Sprintf("ports{>=}%s", peerPorts[0][UUID][0]))
+			if err != nil || len(peerPorts) > 1 {
+				klog.Errorf("failed to list peer switch of %s, %v", peerPorts, err)
+				continue
+			}
+			var aLsw util.LogicalSwitch
+			aLsw.Name = switches[0][Name][0]
+			aRt := lRouters[k]
+			aRt.LogicalSwitch = append(aRt.LogicalSwitch, aLsw)
+			lRouters[k] = aRt
+		}
+	}
+	klog.V(4).Infof("sync over with %s", lRouters)
+
+	slctr := labels.SelectorFromSet(labels.Set{util.VpcExternalAnnotation: "true"})
+	vpcs, err := c.vpcsLister.List(slctr)
+	if err != nil {
+		klog.Errorf("failed to list vpc, %v", err)
+		return
+	}
+	vpcmaps := make(map[string]*v1.Vpc)
+	for _, vpc := range vpcs {
+		vpcmaps[vpc.Name] = vpc
+	}
+	for k, v := range vpcmaps {
+		if _, ok := lRouters[k]; ok {
+			v.Status.Subnets = []string{}
+			for _, asw := range lRouters[k].LogicalSwitch {
+				v.Status.Subnets = append(v.Status.Subnets, asw.Name)
+			}
+			bytes, err := v.Status.Bytes()
+			if err != nil {
+				continue
+			}
+			_, err = c.config.KubeOvnClient.KubeovnV1().Vpcs().Patch(context.Background(), k, types.MergePatchType, bytes, metav1.PatchOptions{}, "status")
+			if err != nil {
+				continue
+			}
+			delete(lRouters, k)
+			klog.V(4).Infof("patch vpc %s", k)
+		} else {
+			opts := metav1.DeleteOptions{}
+			err = c.config.KubeOvnClient.KubeovnV1().Vpcs().Delete(context.Background(), k, opts)
+			if err != nil {
+				continue
+			}
+			klog.V(4).Infof("delete vpc %s ", k)
+		}
+	}
+	if len(lRouters) != 0 {
+		for k, v := range lRouters {
+			vpc := &v1.Vpc{}
+			vpc.Name = k
+			for _, v := range v.LogicalSwitch {
+				vpc.Status.Subnets = append(vpc.Status.Subnets, v.Name)
+			}
+			labels := make(map[string]string)
+			labels[util.VpcExternalAnnotation] = "true"
+			vpc.Labels = labels
+			vpc.Status.DefaultLogicalSwitch = ""
```

Comment:
status should use updateStatus to take effect

Reply:
vpc is first created and then update status now.